target/linux/ath79/files/drivers/mtd/nand/rb4xx_nand.c

```diff
@@ -0,0 +1,269 @@
+// SPDX-License-Identifier: GPL-2.0-only
+/*
+ * NAND flash driver for the MikroTik RouterBoard 4xx series
+ *
+ * Copyright (C) 2008-2011 Gabor Juhos <juhosg@openwrt.org>
+ * Copyright (C) 2008 Imre Kaloz <kaloz@openwrt.org>
+ * Copyright (C) 2015 Bert Vermeulen <bert@biot.com>
+ *
+ * This file was based on the driver for Linux 2.6.22 published by
+ * MikroTik for their RouterBoard 4xx series devices.
+ *
+ * The NAND chip containing the root filesystem is behind an SPI-connected
+ * CPLD. The CPLD also acts as a GPIO expander, and the ALE/CLE/NCE pins are
+ * wired to the CPLD.
+ */
+
+#include <linux/mtd/rawnand.h>
+#include <linux/spi/spi.h>
+#include <linux/platform_device.h>
+#include <linux/gpio/driver.h>
+#include <linux/gpio/consumer.h>
+#include <linux/module.h>
+#include <linux/of.h>
+
+#include <asm/mach-ath79/rb4xx_cpld.h>
+
+struct rb4xx_nand_info {
+	struct nand_chip chip;
+	struct spi_device *spi_dev;
+	struct gpio_desc *gpio_rdy;
+	struct gpio_desc *gpio_ale;
+	struct gpio_desc *gpio_cle;
+	struct gpio_desc *gpio_nce;
+};
+
+static int rb4xx_ooblayout_ecc(struct mtd_info *mtd, int section,
+				struct mtd_oob_region *oobregion)
+{
+	switch (section) {
+	case 0:
+		oobregion->offset = 8;
+		oobregion->length = 3;
+		return 0;
+	case 1:
+		oobregion->offset = 13;
+		oobregion->length = 3;
+		return 0;
+	default:
+		return -ERANGE;
+	}
+}
+
+static int rb4xx_ooblayout_free(struct mtd_info *mtd, int section,
+				struct mtd_oob_region *oobregion)
+{
+	switch (section) {
+	case 0:
+		oobregion->offset = 0;
+		oobregion->length = 4;
+		return 0;
+	case 1:
+		oobregion->offset = 4;
+		oobregion->length = 1;
+		return 0;
+	case 2:
+		oobregion->offset = 6;
+		oobregion->length = 2;
+		return 0;
+	case 3:
+		oobregion->offset = 11;
+		oobregion->length = 2;
+		return 0;
+	default:
+		return -ERANGE;
+	}
+}
+
+static const struct mtd_ooblayout_ops rb4xx_nand_ecclayout_ops = {
+	.ecc = rb4xx_ooblayout_ecc,
+	.free = rb4xx_ooblayout_free,
+};
+
+static int rb4xx_nand_dev_ready(struct mtd_info *mtd)
+{
+	struct nand_chip *chip = mtd->priv;
+	struct rb4xx_nand_info *info = chip->priv;
+	return gpiod_get_value_cansleep(info->gpio_rdy);
+}
+
+static void rb4xx_nand_cmd_ctrl(struct mtd_info *mtd, int cmd,
+				unsigned int ctrl)
+{
+	struct nand_chip *chip = mtd->priv;
+	struct rb4xx_nand_info *info = chip->priv;
+	u8 data = cmd;
+	int ret;
+
+	if (ctrl & NAND_CTRL_CHANGE) {
+		gpiod_set_value_cansleep(info->gpio_cle, !!(ctrl & NAND_CLE));
+		gpiod_set_value_cansleep(info->gpio_ale, !!(ctrl & NAND_ALE));
+		gpiod_set_value_cansleep(info->gpio_nce,  !(ctrl & NAND_NCE));
+	}
+
+	if (cmd != NAND_CMD_NONE) {
+		ret = rb4xx_cpld_write_nand(info->spi_dev, &data, 1);
```

Comment:
You don't need to keep this indirection.
The exported `rb4xx_cpld_*_nand()` functions should take a higher level pointer. As the code currently stands, probably `nand_chip` or `mtd_info` make sense. Then they themselves handle dereferencing the spi data. That's why I said you should be able to remove all spi dependencies to this file.

Of course then the `rb4xx_nand_{read,write}*` functions become simple wrappers, further suggesting that all this code probably belongs to one and the same driver, as you suggested earlier :)